rig/machine_control/scp_connection.py

```diff
@@ -79,65 +125,167 @@ def send_scp(self, buffer_size, x, y, p, cmd, arg1=0, arg2=0, arg3=0,
+        # interface.  This significantly reduces code duplication.
+        # Construct a callable to retain the returned packet for us
+        class Callback(object):
+            def __init__(self):
+                self.packet = None
+
+            def __call__(self, packet):
+                self.packet = SCPPacket.from_bytestring(
+                    packet, n_args=expected_args
+                )
 
-        # Determine how many bytes to listen to on the socket, this should
-        # be the smallest power of two greater than the required size (for
-        # efficiency reasons).
-        max_length = buffer_size + consts.SDP_HEADER_LENGTH
-        receive_length = 1 << 9  # 512 bytes is a reasonable minimum
-        while receive_length < max_length:
-            receive_length <<= 1
+        # Create the packet to send
+        callback = Callback()
+        packets = [
+            scpcall(x, y, p, cmd, arg1, arg2, arg3, data, expected_args,
+                    callback, timeout)
+        ]
 
-        # Repeat until a reply is received or we run out of tries.
-        n_tries = 0
-        while n_tries < self.n_tries:
-            # Transit the packet
-            self.sock.send(packet.bytestring)
-            n_tries += 1
+        # Send the burst
+        self.send_scp_burst(buffer_size, 1, iter(packets))
 
-            try:
-                # Try to receive the returned acknowledgement
-                ack = self.sock.recv(receive_length)
-            except IOError:
-                # There was nothing to receive from the socket
-                continue
-
-            # Convert the possible returned packet into an SCP packet. If
-            # the sequence number matches the expected sequence number then
-            # the acknowledgement has been received.
-            scp = packets.SCPPacket.from_bytestring(ack, n_args=expected_args)
-
-            # Check that the CMD_RC isn't an error
-            if scp.cmd_rc in self.error_codes:
-                raise self.error_codes[scp.cmd_rc](
-                    "Packet with arguments: cmd={}, arg1={}, arg2={}, "
-                    "arg3={}; sent to core ({},{},{}) was bad.".format(
-                        cmd, arg1, arg2, arg3, x, y, p
+        # Return the received packet
+        assert callback.packet is not None
+        return callback.packet
+
+    def send_scp_burst(self, buffer_size, window_size,
+                       parameters_and_callbacks):
+        """Send a burst of SCP packets and call a callback for each returned
+        packet.
+
+        Parameters
+        ----------
+        buffer_size : int
+            Number of bytes held in an SCP buffer by SARK, determines how many
+            bytes will be expected in a socket.
+        window_size : int
+            Number of packets which can be awaiting replies from the SpiNNaker
+            board.
+        parameters_and_callbacks: iterable of :py:class:`.scpcall`
```

Comment:
I prefer parameters and callbacks because that is what's passed in, packets are created internally.